src/main/java/graphql/execution/instrumentation/dataloader/DataLoaderCF.java

```diff
@@ -0,0 +1,99 @@
+package graphql.execution.instrumentation.dataloader;
+
+import graphql.ExperimentalApi;
+import graphql.Internal;
+import graphql.execution.DataLoaderDispatchStrategy;
+import graphql.execution.ExecutionContext;
+import graphql.schema.DataFetchingEnvironment;
+
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CountDownLatch;
+import java.util.function.Supplier;
+
+import static graphql.execution.Execution.EXECUTION_CONTEXT_KEY;
+
+@Internal
+public class DataLoaderCF<T> extends CompletableFuture<T> {
+    final DataFetchingEnvironment dfe;
+    final String dataLoaderName;
+    final Object key;
+    final CompletableFuture<Object> dataLoaderCF;
+
+    volatile CountDownLatch latch;
+
+    public DataLoaderCF(DataFetchingEnvironment dfe, String dataLoaderName, Object key) {
+        this.dfe = dfe;
+        this.dataLoaderName = dataLoaderName;
+        this.key = key;
+        if (dataLoaderName != null) {
```

Comment:
Why take a data loader name such that you need to do dfe.getDataLoaderRegistry().getDataLoader(dataLoaderName) ??

Why not do this lookup on the outside of the creation of the DataLoaderCF class